src/webob/acceptparse.py

```diff
@@ -1641,86 +927,54 @@ def __add__(self, other):
           including any media type parameters, and *extension_params* is a
           ``str`` of the extension parameters segment of the header element,
           starting with the first '``;``'
-        * an :class:`AcceptValidHeader`, :class:`AcceptNoHeader`, or
-          :class:`AcceptInvalidHeader` instance
+        * an :class:`Accept` instance
         * object of any other type that returns a value for ``__str__``
 
-        If `other` is a valid header value or an :class:`AcceptValidHeader`
-        instance, then a new :class:`AcceptValidHeader` instance with the valid
-        header value is returned.
-
-        If `other` is ``None``, an :class:`AcceptNoHeader` instance, an invalid
-        header value, or an :class:`AcceptInvalidHeader` instance, a new
-        :class:`AcceptNoHeader` instance is returned.
+        The rules for adding values to a header are that the values are
+        appended if valid, or discarded. If everything is discarded then an
+        instance representing a missing header is returned.
         """
 
-        if isinstance(other, AcceptValidHeader):
-            return AcceptValidHeader(header_value=other.header_value)
-
-        if isinstance(other, (AcceptNoHeader, AcceptInvalidHeader)):
-            return AcceptNoHeader()
+        other = create_accept_header(other)
+        is_self_valid = self.header_state is HeaderState.Valid
+        is_other_valid = other.header_state is HeaderState.Valid
 
-        return self._add_instance_and_non_accept_type(instance=self, other=other)
+        if is_self_valid:
+            if is_other_valid:
+                if self.header_value == "":
+                    return other
+                if other.header_value == "":
+                    return self
```

Comment:
The docstring states that a new header object is always created, but if `other.header_value == ""`, `self` is returned?